contents/testing/writing-testable-code.md

```diff
@@ -0,0 +1,110 @@
+# Writing Testable Code
+
+Below is a collection of 10 rules that you can follow to write testable code.
+
+## Don't mix object construction with application logic
+
+Because when you are writing tests, you mostly want to isolate some part of the application.
+
+In order to instantiate the class in isolation you have to make sure that the class itself does not instantiate other objects.
+
+So the first rule for writing testable code is do not mix object construction with application logic, instead, you should have two kinds of classes.
+
+First, the **factories**, where all `new` operators reside in, are in charge of constructing objects, and nothing else.
+
+The other, the **application logic classes**, are devoid of `new` operators. They simply ask for them. Thus it makes easier to test the application logic by replacing the real classes for test doubles.
+
+## Don't look for things. Ask for things.
+
+Aka Dependency Injection / Law of Demeter.
+
+The [Law of Demeter](https://en.wikipedia.org/wiki/Law_of_Demeter) can be highlighted as:
+
+* Only talk to your **immediate** friends.
+
+> If you are an Authenticator class don't pass in a path of the configuration information which you read inside the constructor to configure yourself, just ask for the configuration object and let some other class worry about reading the object from the disk.
+
+> Because in your tests you will not want to write a configuration into a disk just so that your object can read it in again.
+
+In this example, you want a configuration object, but instead of directly asking for the object, you *look for* it through a path. It hurts testability because you have to first write configuration information to the path, then pass the path to the Authenticator. Troublesome.
+
+## Don't do actual work in constructors
+
+Because you want to keep yourself away from trouble? Just help yourself and make the constructors do no work.
+
+> A class under tests can have tens of tests. Each test instantiates a slightly different object graph and than applies some stimulus and asserts a response.
+...
+> Any work you do in a constructor needs to successfully navigate through in every test (not just the direct test, but also any related test which tries to instantiate your class indirectly as part of some larger object graph).
+
+In short, the constructor's job is to assign the dependencies to fields. And that's all.
+
+## Avoid Global States
+
+Global states may be tolerable at runtime if you only have one instance of your application.
+
+However, each test is a small instantiation of your application. But the global states persists from one test to another (e.g. a global configuration file).
+
+When tests fail, there is possibility it cannot be reproduced in isolation due to the global states. In other words, order of tests matter, which is bad.
+
+## Avoid Singletons
```

Comment:
If this is going to be a point that is different from "Avoid Global States", maybe you could elaborate more?